pkg/service/did/resolution/resolver.go

```diff
@@ -0,0 +1,108 @@
+package resolution
+
+import (
+	"context"
+
+	didsdk "github.com/TBD54566975/ssi-sdk/did"
+	"github.com/TBD54566975/ssi-sdk/util"
+	"github.com/pkg/errors"
+	"github.com/sirupsen/logrus"
+
+	didint "github.com/tbd54566975/ssi-service/internal/did"
+	utilint "github.com/tbd54566975/ssi-service/internal/util"
+)
+
+type Resolver interface {
+	Resolve(ctx context.Context, did string, opts ...didsdk.ResolutionOptions) (*didsdk.ResolutionResult, error)
+}
+
+// ServiceResolver is a resolver that can resolve DIDs using a combination of local and universal resolvers.
+type ServiceResolver struct {
+	resolutionMethods []string
+	hr                Resolver
+	lr                *localResolver
+	ur                *universalResolver
+}
+
+// NewServiceResolver creates a new ServiceResolver instance which can resolve DIDs using a combination of local and
+// universal resolvers.
+func NewServiceResolver(handlerResolver Resolver, resolutionMethods []string, universalResolverURL string) (*ServiceResolver, error) {
+	if len(resolutionMethods) == 0 {
+		return nil, errors.New("no resolution methods configured")
+	}
+
+	// instantiate sdk resolver
+	sdkResolver, err := didint.BuildResolver(resolutionMethods)
+	if err != nil {
+		return nil, errors.Wrap(err, "instantiating SDK DID resolver")
+	}
+	lr, err := newLocalResolver(sdkResolver)
+	if err != nil {
+		return nil, errors.Wrap(err, "instantiating local resolver")
+	}
+
+	// instantiate universal resolver
+	var ur *universalResolver
+	if universalResolverURL != "" {
+		ur, err = newUniversalResolver(universalResolverURL)
+		if err != nil {
+			return nil, errors.Wrap(err, "instantiating universal resolver")
+		}
+	}
+
+	return &ServiceResolver{
+		resolutionMethods: resolutionMethods,
+		hr:                handlerResolver,
+		lr:                lr,
+		ur:                ur,
+	}, nil
+}
+
+// Resolve resolves a DID using a combination of local and universal resolvers. The ordering is as follows:
+// 1. Try to resolve with the handlers we have, wrapping the resulting DID in resolution result
```

Comment:
For what use-cases do we need to support resolution with our handlers?

Reply:
I am imagining the case of something like did:web or ion - any remote resolution - where we do not need to call out to a service endpoint if we have a local copy. this raises the question of staleness, where we can add a resolution option for `forceRemote`

Reply:
Should we move away from using storage as a DID document cache?

Or if we're using it as a cache, then I would recommend we keep that in memory (or in a redis implementation that doesn't need to guarantee durability).

Why I'm more favorable to removing is because caches are used as a performance optimization, and can lead to staleness bugs. Since we haven't diagnosed performance bottlenecks, I don't think they're worth the complexity. I've already seen some confusion pop up from how the DIDAPI endpoint works.

What are your thoughts? If you agree, mind creating a tracking issue?

Reply:
I'm fine with killing caches, but we should definitely keep storing DID Documents in storage not as a cache

Reply:
https://github.com/TBD54566975/ssi-service/issues/361